aggregator/src/aggregator.rs

```diff
@@ -478,24 +575,169 @@ impl<C: Clock> Aggregator<C> {
+                let task = task.clone();
+                Box::pin(async move { tx.put_task(&task.into()).await })
+            })
+            .await
+            .or_else(|error| -> Result<(), Error> {
+                match error {
+                    // If the task is already in the datastore, then some other request or aggregator
+                    // replica beat us to inserting it. They _should_ have inserted all the same parameters
+                    // as we would have, so we can proceed as normal.
+                    DatastoreError::MutationTargetAlreadyExists => {
+                        warn!(
+                            ?task_id,
+                            ?error,
+                            "taskprov: went to insert task into db, but it already exists"
+                        );
+                        Ok(())
+                    }
+                    error => Err(error.into()),
+                }
+            })?;
+
+        info!(?task, ?peer_aggregator, "taskprov: opted into new task");
+        Ok(())
+    }
+
+    /// Validate and authorize a taskprov request. Returns values necessary for determining whether
+    /// we can opt into the task. This function might return an opt-out error for conditions that
+    /// are relevant for all DAP workflows (e.g. task expiration).
+    #[tracing::instrument(skip(self, aggregator_auth_token), err)]
+    async fn taskprov_authorize_request(
+        &self,
+        peer_role: &Role,
+        task_id: &TaskId,
+        task_config: &TaskConfig,
+        aggregator_auth_token: Option<&AuthenticationToken>,
+    ) -> Result<(&PeerAggregator, Vec<Url>), Error> {
+        let aggregator_urls = task_config
+            .aggregator_endpoints()
+            .iter()
+            .map(|url| url.try_into())
+            .collect::<Result<Vec<Url>, _>>()?;
+        if aggregator_urls.len() < 2 {
+            return Err(Error::UnrecognizedMessage(
+                Some(*task_id),
+                "taskprov configuration is missing one or both aggregators",
+            ));
+        }
+        let peer_aggregator_url = &aggregator_urls[peer_role.index().unwrap()];
+
+        let peer_aggregator = self
+            .peer_aggregators
+            .get(peer_aggregator_url, peer_role)
+            .ok_or(Error::InvalidTask(
+                *task_id,
+                OptOutReason::NoSuchPeer(*peer_role),
+            ))?;
+
+        if !aggregator_auth_token
+            .map(|t| peer_aggregator.check_aggregator_auth_token(t))
+            .unwrap_or(false)
+        {
+            return Err(Error::UnauthorizedRequest(*task_id));
+        }
+
+        if self.clock.now() > *task_config.task_expiration() {
+            return Err(Error::InvalidTask(*task_id, OptOutReason::TaskExpired));
         }
+
+        info!(
+            ?task_id,
+            ?task_config,
+            ?peer_aggregator,
+            "taskprov: authorized request"
+        );
```

Comment:
I don't think we should log every time we successfully authorize a request.

Reply:
I've demoted it to `debug` level.